File: Lab06/C/hetmani.c
#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>

// bije_wiersz -

void Ustaw(int i,int n, int* position,bool* bije_kolumne, bool* bije_przek1, bool* bije_przek2, int* licznik){

    for (int j = 0; j < n;j++){

        int index_przek1 = i + j;

        int index_przek2 = i - j + (n - 1);

        if (!(bije_kolumne[j] || bije_przek1[index_przek1] || bije_przek2[index_przek2])){
            position[i] = j;
            bije_kolumne[j] = true;
            bije_przek1[index_przek1] = true;
            bije_przek2[index_przek2] = true;

            if (i < n - 1){
                Ustaw(i+1,n, position, bije_kolumne, bije_przek1, bije_przek2, licznik);
            }
            else{
                (*licznik)++; // w nawiasach aby zwiększyć wartość zmiennej a nie adres
                for (int k = 0; k < n; k++){
                    printf("%d ", position[k] + 1);
                }
                printf("\n");
            }
            position[i] = 0;
            bije_kolumne[j] = false;
            bije_przek1[index_przek1] = false;
            bije_przek2[index_przek2] = false; 
        }
    }

}

int hetmani(int n){
    int *position = calloc(n, sizeof(int));
    bool *bije_kolumne = calloc(n, sizeof(bool));
    bool *bije_przek1 = calloc(2*n, sizeof(bool));
    bool *bije_przek2 = calloc(2*n, sizeof(bool));
    int licznik = 0;
    Ustaw(0, n, position, bije_kolumne, bije_przek1, bije_przek2, &licznik);
    free(position);
    free(bije_kolumne);
    free(bije_przek1);
    free(bije_przek2);
    return licznik;
}

int main(int argc, char* argv[]){
    if (argc != 2){
        printf("Zła liczba argumentów\n");
        return 1;
    }

    int n = atoi(argv[1]);
    
    if (n <= 1){
        printf("Błędny argument\n");
        return 1;
    }
    printf("%d", hetmani(n));
    printf("\n");
    return 0;
}

File: semestr1/Lab06/Python/hetmani.py
import sys



    


def hetmani(n):
    
    licznik = 0

    def Ustaw(i, n, position, bije_kolumne, bije_przek1, bije_przek2):

        nonlocal licznik

        for j in range(n):
            index_przek1 = i + j

            index_przek2 = i - j + (n-1)

            if (not (bije_kolumne[j] or bije_przek1[index_przek1] or bije_przek2[index_przek2])):
                position[i] = j
                bije_kolumne[j] = True
                bije_przek1[index_przek1] = True
                bije_przek2[index_przek2] = True

                if (i < n - 1):
                    Ustaw(i+1,n, position, bije_kolumne, bije_przek1, bije_przek2)
                else:

                    for k in range(n):
                        print(f"{position[k] + 1} ", end="")
                    print("")
                    licznik += 1
                
                position[i] = 0
                bije_kolumne[j] = False
                bije_przek1[index_przek1] = False
                bije_przek2[index_przek2] = False

    position = [0] * n
    bije_kolumne = [False] * n
    bije_przek1 = [False] * (2*n)
    bije_przek2 = [False] * (2*n)

    Ustaw(0, n, position, bije_kolumne, bije_przek1, bije_przek2)
    return licznik



def main():
    if len(sys.argv) != 2:
        print("Zła liczba argumentów")
        return
    
    n = int(sys.argv[1])

    if (n < 1):
        print("Błędny argument")
        return
    
    print(f"{hetmani(n)}")


if (__name__ == "__main__"):
    main()